### PR TITLE
GFORMS-1451 Implemented BARS check

### PR DESCRIPTION
**Pull request by girishkamat**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import uk.gov.hmrc.gform.core.Opt
+import uk.gov.hmrc.gform.core.parsers.{ BasicParsers, ValueParser }
+import uk.gov.hmrc.gform.exceptions.UnexpectedState
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.{ Expr, JsonUtils, OFormatWithTemplateReadFallback }
+
+import scala.util.Try
+import scala.util.matching.Regex
+
+sealed trait DataRetrieveType
+trait ValidateBankType extends DataRetrieveType
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala:31]: What for is this?

File: app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                            } yield ValidateBank(DataRetrieveId(idValue), sortCodeExpr, accountNumberExpr)
+                          case other => Left(UnexpectedState(s"'type' value $other not recognized"))
+                        }
+      } yield dataRetrieve).fold(e => JsError(e.error), r => JsSuccess(r))
+  }
+
+  def opt[T](jsValue: JsValue, path: String)(implicit r: Reads[T]): Opt[T] =
+    jsValue \ path match {
+      case JsDefined(json) =>
+        Try(json.as[T])
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala:102]: There is `validate` method for checking if json can be converted to  type `T`

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala (hunk 1 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala]: Removed as it was unused

File: app/uk/gov/hmrc/gform/models/constraints/ReferenceInfo.scala [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 package uk.gov.hmrc.gform.models.constraints
 
-import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.{ DataRetrieveCtx, _ }
```
[the remark this answers is not in the document]
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/models/constraints/ReferenceInfo.scala]: Removed

File: app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala (hunk 2 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala]: Replaced with validate method

**Review by josef-vlach — APPROVED** [review]

**Pull request merged**